### PR TITLE
ARROW-1164: C++: Templated functions need ARROW_EXPORT instead of ARROW_TEMPLATE_EXPORT

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

File: cpp/src/arrow/util/visibility.h [hunk 1]
```diff
@@ -52,7 +52,7 @@
 
 // This is a complicated topic, some reading on it:
 // http://www.codesynthesis.com/~boris/blog/2010/01/18/dll-export-cxx-templates/
-#if defined(_MSC_VER) || defined(__clang__)
+#if defined(_MSC_VER)
```
Comment (xhochy) [inline cpp/src/arrow/util/visibility.h]: Reverting this as I accidentially though this was the fix.

File: cpp/src/arrow/python/helpers.cc [hunk 2]
```diff
@@ -90,11 +90,11 @@ Status PythonDecimalToArrowDecimal(
...
 
-template Status ARROW_TEMPLATE_EXPORT PythonDecimalToArrowDecimal(
+template Status ARROW_EXPORT PythonDecimalToArrowDecimal(
     PyObject* python_decimal, decimal::Decimal32* arrow_decimal);
-template Status ARROW_TEMPLATE_EXPORT PythonDecimalToArrowDecimal(
+template Status ARROW_EXPORT PythonDecimalToArrowDecimal(
     PyObject* python_decimal, decimal::Decimal64* arrow_decimal);
-template Status ARROW_TEMPLATE_EXPORT PythonDecimalToArrowDecimal(
+template Status ARROW_EXPORT PythonDecimalToArrowDecimal(
     PyObject* python_decimal, decimal::Decimal128* arrow_decimal);
```
Comment (wesm) [inline cpp/src/arrow/python/helpers.cc]: Could this also be fixed by adding extern template declarations in helper.h, like we've done elsewhere?

Comment (xhochy) [inline cpp/src/arrow/python/helpers.cc]: I'm not sure why we would need another template here. In the end, this is simply a function implementation.

Comment (wesm) [inline cpp/src/arrow/python/helpers.cc]: Yeah maybe just drop the template altogether and have the 3 different function signatures. @cpcloud?

Comment (cpcloud) [inline cpp/src/arrow/python/helpers.cc]: @xhochy @wesm I can refactor to not use templates. I'll open a PR up shortly.

Comment (cpcloud) [inline cpp/src/arrow/python/helpers.cc]: https://github.com/apache/arrow/pull/794

**Pull request closed without merging**